test/integration/network/services_test.go

```diff
@@ -0,0 +1,130 @@
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	netutils "k8s.io/utils/net"
+
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+// TestServicesFinalizers tests that services participate in the object
+// deletion when using finalizers
+func TestServicesFinalizers(t *testing.T) {
+
+	serviceCIDR := "10.0.0.0/16"
+
+	cfg := framework.NewIntegrationTestControlPlaneConfig()
+	_, cidr, err := netutils.ParseCIDRSloppy(serviceCIDR)
+	if err != nil {
+		t.Fatalf("bad cidr: %v", err)
+	}
+	cfg.ExtraConfig.ServiceIPRange = *cidr
+	cfg.ExtraConfig.RepairServicesInterval = 5 * time.Second
+	_, s, closeFn := framework.RunAnAPIServer(cfg)
+	defer closeFn()
+
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL})
+
+	// verify client is working
+	if err := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+		_, err = client.CoreV1().Endpoints(metav1.NamespaceDefault).Get(context.TODO(), "kubernetes", metav1.GetOptions{})
+		if err != nil {
+			t.Logf("error fetching endpoints: %v", err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Errorf("server without enabled endpoints failed to register: %v", err)
+	}
+
+	// Create a NodePort service with one finalizer
+	svcNodePort := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:       "svc",
+			Finalizers: []string{"foo.bar/some-finalizer"},
+		},
+		Spec: v1.ServiceSpec{
+			ClusterIP: "10.0.0.20",
+			Ports: []v1.ServicePort{{
+				Port:       8443,
+				NodePort:   30443,
+				TargetPort: intstr.FromInt(8443),
+				Protocol:   v1.ProtocolTCP,
+			}},
+			Type: v1.ServiceTypeNodePort,
+		},
+	}
+
+	// Create service
+	if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), &svcNodePort, metav1.CreateOptions{}); err != nil {
+		t.Errorf("unexpected error creating service: %v", err)
+	}
+	t.Logf("Created service: %s", svcNodePort.Name)
+
+	// Check the service has been created correctly
+	svc, err := client.CoreV1().Services(metav1.NamespaceDefault).Get(context.TODO(), svcNodePort.Name, metav1.GetOptions{})
+	if err != nil || svc.Spec.ClusterIP != "10.0.0.20" {
+		t.Errorf("created service is not correct: %v", err)
+	}
+	t.Logf("Service created successfully: %v", svc)
+
+	// Delete service
+	if err := client.CoreV1().Services(metav1.NamespaceDefault).Delete(context.TODO(), svcNodePort.Name, metav1.DeleteOptions{}); err != nil {
+		t.Errorf("unexpected error deleting service: %v", err)
+	}
+	t.Logf("Deleted service: %s", svcNodePort.Name)
+
+	// wait for the repair loop to recover the deleted resources
+	time.Sleep(cfg.ExtraConfig.RepairServicesInterval)
```

Comment:
This feels like it could flake - should you 2x it ?

Reply:
I've added 1 second and it works fine, we should not make these takes take too much time, it takes 15 sec now 
```
real    0m15,014s
```

```
stress ./network.test 
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 12 runs so far, 0 failures
20s: 24 runs so far, 0 failures
25s: 24 runs so far, 0 failures
30s: 35 runs so far, 0 failures
35s: 48 runs so far, 0 failures
40s: 50 runs so far, 0 failures
45s: 58 runs so far, 0 failures
50s: 72 runs so far, 0 failures
55s: 74 runs so far, 0 failures
1m0s: 86 runs so far, 0 failures
```

Reply:
You could make the interval 2 or 3 seconds?